File: Julia/coll_arb.jl
# CSC 330
# Assignmant 3 - Collatz Conjecture
#
# Author: Devin McBryde
#
#
#

#stores the top ten highest values and the steps to 1
maxValues = zeros(Int64, 2, 10)

#the location of the minimum value in the array
minVal = 1

#recursive function that return the number of steps needed to return
#	to one for the input number
function collatzStep(a::Int64)::Int64

	#if convert(Bool,(a & 1))
	if a%2 == 1
		a = a*3 + 1
	else
		a = a>>1
	end

	return a

end 

#maxValues[1,4] = 1

i = parse(BigInt,"1000000000000000")

#Main loop that goes through all values between 2 and 5000000000
#	Top value has the L suffix since literals are interpreted as integers
while i > 999999999000000

	alreadyexists = 0

	#reset the next twp values for the new number
	#col holds the value of the iterated number
	col::Int64 = i
	#stepCount tracks the number of iterations total	
	stepCount::Int64 = 0

	#Here we iterate values until they hit 1
	while col != 1

		#next two lines use the function, however the program then
		#	runs much slower

		#stepCount += 1
		#col = collatzStep(col)

		#This if statement will perform the operations the collatz sequence

		if col%2 == 1

			#If the number is odd the next number must be even
			#	therefore we can avoid logic checks by 
			#	performing both operations and adding 2
			#	to the count
			col = (col*3 + 1)>>1
			stepCount += 2
		else
			#If the number is even we divide by two and add one to count
			col = col>>1
			stepCount += 1
		end

	end

	#Here we avoid having a value with a duplicate number of steps using the boolean flag
	for j = 1:10
		#We check if our count has been recorded already
		if stepCount == maxValues[1,j]
			alreadyexists = 1
		end
	end

	#Here we check if the count is larger than the smallest count recorded and add if it is
	if stepCount > maxValues[1,minVal] && !(alreadyexists == 1)

		#here we replace the value of the smallest count
		for j = 1:10

			if j == minVal 

				maxValues[1,j] = stepCount
				maxValues[2,j] = i

			end

		end

		#we now reset the minVal to look for the new lowest count value

		minVal = 1

		#search for the smallest count size in maxValues
		for j = 1:10

			if maxValues[1,j] < maxValues[1,minVal]
				minVal = j
			end

		end

	end

	i = i - 1

end

#Now we perform a basic selection sort on the step count before printing
for i = 1:9

	minValue::Int64 = maxValues[1,i]
	minColNum::Int64  = maxValues[2,i]
	minLocale = i
	
	for j = (i+1):10
	
		if minValue < maxValues[1,j]
			minValue = maxValues[1,j]
			minColNum = maxValues[2,j]
			minLocale = j
		end
	
	end
	
	tempVal::Int64  = maxValues[1,i]
	tempNum::Int64  = maxValues[2,i]
	
	maxValues[1,i] = minValue
	maxValues[1,minLocale] = tempVal

	maxValues[2,i] = minColNum
	maxValues[2,minLocale] = tempNum
	
end

#print the maxValues array
for i = 1:10
	println("Values: ", maxValues[2,i]," Steps Taken: ", maxValues[1,i])
end

File: Java/Coll_arb.java
/* CSC 330
 * Assignmant 3 - Collatz Conjecture
 *
 * Author: Devin McBryde
 *
 *
 */

import java.math.BigInteger;

public class Coll_arb{

	public static void main(String[] args){	

		//stores the top ten highest values and the steps to 1
		BigInteger maxValues[][] = new BigInteger[2][10];

		for (int i = 0; i < 10; i++){
			maxValues[0][i] = BigInteger.ZERO;
			maxValues[1][i] = BigInteger.ZERO;
		}
	
		//the location of the minimum value in the array
		int minVal = 0;	

		BigInteger test = new BigInteger("5");

		if(test.testBit(2)){
			//System.out.println("true");
		}

		//Main loop that goes through all values between 2 and 5000000000
		//	Top value has the L suffix since literals are interpreted as integers	
		for (BigInteger i = new BigInteger("1000000000000000"); i.compareTo(BigInteger.valueOf(2)) == 1; i = i.subtract(BigInteger.ONE) ){

			boolean alreadyexists = false;

			//reset the next two values for the new number
			//col holds the value of the iterated number
			BigInteger col = i;
			//count tracks the number of iterations total
			BigInteger count = new BigInteger("0");
	
			//Here we iterate values until they hit 1
			while(col.compareTo(BigInteger.valueOf(1)) != 0){
		
				//next two lines use the function, however the program then
				//	runs much slower
				//count += 1;
				//col = collatzStep(col);
				
				//This if statement will perform the operations the collatz sequence
				if( col.testBit(0) ){

					//If the number is odd the next number must be even
					//	therefore we can avoid logic checks by 
					//	performing both operations and adding 2
					//	to the count
					col = col.multiply(BigInteger.valueOf(3)).add(BigInteger.valueOf(1));
					count = count.add(BigInteger.ONE);
				}else{
					//If the number is even we divide by two and add one to count
					col = col.divide(BigInteger.valueOf(2));
					count = count.add(BigInteger.ONE);
				}			
				
				//check for some overflow
				//if(col < 0){
				//	System.out.println("overflow occured");
				//	break;
				//}			
			}	

			//Here we avoid having a value with a duplicate number of steps using the boolean flag
			for(int j = 0; j < 10; j++){
				//We check if our count has been recorded already
				if (count.compareTo(maxValues[0][j]) == 0 ){
					alreadyexists = true;
				}
			}

			//Here we check if the count is larger than the smallest count recorded and add if it is
			if(count.compareTo(maxValues[0][minVal]) == 1 && !alreadyexists){

				//here we replace the value of the smallest count
				for(int j = 0; j < 10; j++){
					 
					if (j == minVal){
						maxValues[0][j] = count;
						maxValues[1][j] = i;
					}
				}

				//we now reset the minVal to look for the new lowest count value
				minVal = 0;
				
				//search for the smallest count size in maxValues
				for(int j = 0; j < 10; j++){
					if (maxValues[0][j].compareTo(maxValues[0][minVal]) == -1){
						minVal = j;
					}
				}

			}
			
	
		}
	
		
		//Now we perform a basic selection sort on the step count before printing
		for(int i = 0; i < 9; i++){

			BigInteger minValue = maxValues[0][i];
			BigInteger minColNum = maxValues[1][i];
			int minLocale = i;
	
			for(int j = i+1; j < 10; j++){
	
				if(minValue.compareTo(maxValues[0][j]) == -1){
					minValue = maxValues[0][j];
					minColNum = maxValues[1][j];
					minLocale = j;
				}
	
			}
	
			BigInteger tempVal = maxValues[0][i];
			BigInteger tempNum = maxValues[1][i];
	
			maxValues[0][i] = minValue;
			maxValues[0][minLocale] = tempVal;
	
			maxValues[1][i] = minColNum;
			maxValues[1][minLocale] = tempNum;
	
		}

		//print the maxValues array
		for(int i = 0; i < 10; i++){
	
			System.out.println("Value: " + maxValues[1][i] + " Steps Taken: " + maxValues[0][i]);
	
		}


	}

	//Function that will perform a collatz step
	private static long collatzStep(long input){
	
		if(input%2 == 1){
			input = input*3 +1;
			}else{
			input = input/2;
		}
	
		return input;
	
	}
}
